client/src/globalState.tsx

```diff
@@ -328,6 +338,77 @@ export const GlobalStateProvider: React.FC<GlobalStateProviderProps> = ({
     });
   }, []);
 
+  const calculateBoundedPositionTranslation = (
+    positionTranslation: ControlPosition
+  ) => {
+    const isTooFarRight = positionTranslation.x > 0;
+    const isTooLow = positionTranslation.y > 0;
+
+    const viewportWidth = document.documentElement.clientWidth;
+    const viewportHeight = document.documentElement.clientHeight;
+
+    const isTooFarLeft = viewportWidth + positionTranslation.x < floatySize;
+    const isTooHigh = viewportHeight + positionTranslation.y < floatySize;
+    return {
+      x: isTooFarLeft
+        ? 10 + floatySize - viewportWidth
+        : isTooFarRight
+        ? -10
+        : positionTranslation.x,
+      y: isTooHigh
+        ? 10 + floatySize - viewportHeight
+        : isTooLow
+        ? -10
+        : positionTranslation.y,
+    };
+  };
+
+  const [
+    explicitPositionTranslation,
+    setExplicitPositionTranslationState,
+  ] = useState<ControlPosition>({ x: 0, y: 0 });
+  const setExplicitPositionTranslation = (newPosition: ControlPosition) => {
+    setExplicitPositionTranslationState(newPosition);
+    window.localStorage.setItem(
+      LOCAL_STORAGE_KEY_EXPLICIT_POSITION,
+      JSON.stringify(newPosition)
+    );
+  };
+
+  const [
+    boundedPositionTranslation,
+    setBoundedPositionTranslation,
+  ] = useState<ControlPosition>({ x: 0, y: 0 });
+  // position translation must be passed in rather than using explicitPositionTranslation to avoid a rerender briefly in the old position
+  const updateBoundedPositionTranslation = (
+    positionTranslation: ControlPosition
+  ) =>
+    setBoundedPositionTranslation(
+      calculateBoundedPositionTranslation(positionTranslation)
+    );
+
+  const [lastResized, setLastResized] = useState<number>(0);
+
+  useEffect(() => {
+    updateBoundedPositionTranslation(explicitPositionTranslation);
+  }, [lastResized]);
+
+  const resizeCompleteHandler = useDebounce(
+    () => setLastResized(Date.now()),
+    250
+  );
```

Comment:
adding this bit of state (`lastResized`) was the only way I could make it use the latest value for `explicitPositionTranslation` when calling `updateBoundedPositionTranslation` - if I called that directly in the debounce handler it would always use the initial value for `explicitPositionTranslation`, which makes sense, but there's probably a more idiomatic/nicer way to achieve this in React.

Reply:
I think it's that the useEffect adding the resize event handler never removes the handler, and doesn't declare the resizeCompleteHandler as a dependency. This works for me ->

```tsx
const resizeCompleteHandler = useCallback(() => {
  updateBoundedPositionTranslation(explicitPositionTranslation);
}, [explicitPositionTranslation]);

useEffect(() => {
  const savedExplicitPositionTranslation = JSON.parse(
    window.localStorage.getItem(LOCAL_STORAGE_KEY_EXPLICIT_POSITION) ||
      JSON.stringify({ x: 0 - right, y: 0 - bottom })
  );
  setExplicitPositionTranslation(savedExplicitPositionTranslation);
  updateBoundedPositionTranslation(savedExplicitPositionTranslation);

  window.addEventListener("resize", resizeCompleteHandler);
  return window.removeEventListener("resize", resizeCompleteHandler);
}, [resizeCompleteHandler]);
```

Reply:
after some pairing we settled on a more idomatic approach, making use of `useCallback` rather than `useState` but ensuring we don't add/remove the resize listener unnecessarily frequently - see https://github.com/guardian/pinboard/pull/148/commits/19246ac6983f527b6b0aee2efef4e0d92e12ea03